pirouette-plutusir/src/Language/Pirouette/PlutusIR/SMT.hs

```diff
@@ -0,0 +1,297 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PatternSynonyms #-}
+
+module Language.Pirouette.PlutusIR.SMT where
+
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Base64 as B64
+import Data.Text (Text)
+import qualified Data.Text.Encoding as T
+import Language.Pirouette.PlutusIR.Syntax
+import Pirouette.Monad
+import Pirouette.SMT.Base
+import Pirouette.SMT.Constraints
+import Pirouette.Symbolic.Eval.Types
+import Pirouette.Term.Syntax.Base
+import Pirouette.Term.Syntax.SystemF as SystemF
+import qualified PureSMT
+
+-- See https://github.com/input-output-hk/plutus/blob/master/plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs
+
+instance LanguageSMT PlutusIR where
+  translateBuiltinType = trPIRType
+
+  -- translateBuiltinTerm = error "translateBuiltinTerm (t :: BuiltinTerms PlutusIR): not yet impl"
+  translateBuiltinTerm = trPIRFun
+  translateConstant = trPIRConstant
+
+  isStuckBuiltin e
+    | termIsConstant e = True
+    | Just _ <- termIsMeta e = True
+  isStuckBuiltin (App (Free (Builtin op)) args)
+    | op `elem` plutusIRBasicOps || op `elem` plutusIRBasicRels,
+      all isArg args -- this ensures that we only have TermArg
+      =
+      let args' = map (\(TermArg a) -> a) args
+       in all isStuckBuiltin args' && not (all termIsConstant args')
+  isStuckBuiltin _ = False
+
+trPIRType :: PIRBuiltinType -> PureSMT.SExpr
+trPIRType PIRTypeInteger = PureSMT.tInt
+trPIRType PIRTypeString = PureSMT.tString
+trPIRType PIRTypeByteString = PureSMT.tString
+
+-- TODO Implement remaining constants
+trPIRConstant :: PIRConstant -> PureSMT.SExpr
+trPIRConstant (PIRConstInteger n) = PureSMT.int n
+trPIRConstant (PIRConstByteString bs) = PureSMT.text $ T.decodeUtf8 $ B64.encode bs
```

Comment:
I think it should be "encode" because of the allowable characters for SMTLIB v2 symbols: by calling `B64.encode` we ensure that the test will only contain base64 characters.